File: go/appencryption/pkg/persistence/sql.go
package persistence

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"regexp"
	"strconv"
	"time"

	"github.com/pkg/errors"
	"github.com/rcrowley/go-metrics"

	"github.com/godaddy/asherah/go/appencryption"
)

const (
	defaultLoadKeyQuery    = "SELECT key_record FROM encryption_key WHERE id = ? AND created = ?"
	defaultStoreKeyQuery   = "INSERT INTO encryption_key (id, created, key_record) VALUES (?, ?, ?)"
	defaultLoadLatestQuery = "SELECT key_record from encryption_key WHERE id = ? ORDER BY created DESC LIMIT 1"
)

var (
	// Verify SQLMetastore implements the Metastore interface.
	_ appencryption.Metastore = (*SQLMetastore)(nil)

	storeSQLTimer      = metrics.GetOrRegisterTimer(fmt.Sprintf("%s.metastore.sql.store", appencryption.MetricsPrefix), nil)
	loadSQLTimer       = metrics.GetOrRegisterTimer(fmt.Sprintf("%s.metastore.sql.load", appencryption.MetricsPrefix), nil)
	loadLatestSQLTimer = metrics.GetOrRegisterTimer(fmt.Sprintf("%s.metastore.sql.loadlatest", appencryption.MetricsPrefix), nil)
)

// SQLMetastoreDBType identifies a specific database/sql driver.
type SQLMetastoreDBType string

const (
	Postgres SQLMetastoreDBType = "postgres"
	Oracle   SQLMetastoreDBType = "oracle"
	MySQL    SQLMetastoreDBType = "mysql"

	DefaultDBType = MySQL
)

var qrx = regexp.MustCompile(`\?`)

// q converts "?" characters to $1, $2, $n on postgres, :1, :2, :n on Oracle.
//
// This function is based on a function of the same name found in the Go
// sql test project: https://github.com/bradfitz/go-sql-test.
func (t SQLMetastoreDBType) q(sql string) string {
	var pref string

	//nolint:exhaustive
	switch t {
	case Postgres:
		pref = "$"
	case Oracle:
		pref = ":"
	default:
		return sql
	}
	n := 0
	return qrx.ReplaceAllStringFunc(sql, func(string) string {
		n++
		return pref + strconv.Itoa(n)
	})
}

// SQLMetastoreOption is used to configure additional options in a SQLMetastore.
type SQLMetastoreOption func(*SQLMetastore)

// WithSQLMetastoreDBType configures the SQLMetastore for use with the specified
// family of database/sql drivers such as Postgres, Oracle, or MySQL (default).
func WithSQLMetastoreDBType(t SQLMetastoreDBType) SQLMetastoreOption {
	return func(s *SQLMetastore) {
		s.dbType = t
		s.loadKeyQuery = t.q(s.loadKeyQuery)
		s.storeKeyQuery = t.q(s.storeKeyQuery)
		s.loadLatestQuery = t.q(s.loadLatestQuery)
	}
}

// SQLMetastore implements the Metastore interface for a RDBMS metastore.
//
// See https://github.com/godaddy/asherah/blob/master/docs/Metastore.md#rdbms for the
// required table structure and other relevant information.
type SQLMetastore struct {
	db *sql.DB

	dbType          SQLMetastoreDBType
	loadKeyQuery    string
	storeKeyQuery   string
	loadLatestQuery string
}

// NewSQLMetastore returns a new SQLMetastore with the provided policy and sql connection.
func NewSQLMetastore(dbHandle *sql.DB, opts ...SQLMetastoreOption) *SQLMetastore {
	metastore := &SQLMetastore{
		db: dbHandle,

		dbType:          DefaultDBType,
		loadKeyQuery:    defaultLoadKeyQuery,
		storeKeyQuery:   defaultStoreKeyQuery,
		loadLatestQuery: defaultLoadLatestQuery,
	}

	for _, opt := range opts {
		opt(metastore)
	}

	return metastore
}

type scanner interface {
	Scan(v ...interface{}) error
}

func parseEnvelope(s scanner) (*appencryption.EnvelopeKeyRecord, error) {
	var keyRecordString string

	if err := s.Scan(&keyRecordString); err != nil {
		if errors.Is(err, sql.ErrNoRows) {
			return nil, nil
		}

		return nil, errors.Wrap(err, "error from scanner")
	}

	var keyRecord *appencryption.EnvelopeKeyRecord

	if err := json.Unmarshal([]byte(keyRecordString), &keyRecord); err != nil {
		return nil, errors.Wrap(err, "unable to unmarshal key")
	}

	return keyRecord, nil
}

// Load returns the key matching the id and created timestamp provided. The envelope
// will be nil if it does not exist in the metastore.
func (s *SQLMetastore) Load(ctx context.Context, keyID string, created int64) (*appencryption.EnvelopeKeyRecord, error) {
	defer loadSQLTimer.UpdateSince(time.Now())

	t := time.Unix(created, 0)

	return parseEnvelope(s.db.QueryRowContext(ctx, s.loadKeyQuery, keyID, t))
}

// LoadLatest returns the newest record matching the ID.
func (s *SQLMetastore) LoadLatest(ctx context.Context, keyID string) (*appencryption.EnvelopeKeyRecord, error) {
	defer loadLatestSQLTimer.UpdateSince(time.Now())

	return parseEnvelope(s.db.QueryRowContext(ctx, s.loadLatestQuery, keyID))
}

// Store attempts to insert the key into the metastore if one is not
// already present. If a key exists, the method will return false. If
// one is not present, the value will be inserted and we return true.
// Note that as of this writing, the Go sql package doesn't expose a
// way to detect duplicate keys, so they are treated similarly to all
// errors that may happen in the insert.
func (s *SQLMetastore) Store(ctx context.Context, keyID string, created int64, envelope *appencryption.EnvelopeKeyRecord) (bool, error) {
	defer storeSQLTimer.UpdateSince(time.Now())

	bytes, err := json.Marshal(envelope)
	if err != nil {
		return false, errors.Wrap(err, "error marshaling envelope")
	}

	createdAt := time.Unix(created, 0)

	if _, err := s.db.ExecContext(ctx, s.storeKeyQuery, keyID, createdAt, string(bytes)); err != nil {
		// Go sql package does not provide a specific integrity violation error for duplicate detection
		// at this time, so it's treated similar to other errors to avoid error parsing.
		// The caller is left to assume any false/error return value may be a duplicate.
		return false, errors.Wrapf(err, "error storing key: %s, %d", keyID, created)
	}

	return true, nil
}

File: rust/appencryption/src/persistence/sql.rs
use crate::envelope::EnvelopeKeyRecord;
use crate::error::{Error, Result};
use crate::Metastore;

use async_trait::async_trait;
use chrono::{DateTime, TimeZone, Utc};
use metrics::histogram;
use regex::Regex;
use std::sync::Arc;
use std::time::Instant;

// Default SQL queries for different operations
const DEFAULT_LOAD_KEY_QUERY: &str =
    "SELECT key_record FROM encryption_key WHERE id = ? AND created = ?";
const DEFAULT_STORE_KEY_QUERY: &str =
    "INSERT INTO encryption_key (id, created, key_record) VALUES (?, ?, ?)";
const DEFAULT_LOAD_LATEST_QUERY: &str =
    "SELECT key_record from encryption_key WHERE id = ? ORDER BY created DESC LIMIT 1";

/// Database type for SQL metastore
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SqlMetastoreDbType {
    /// MySQL database
    MySql,

    /// PostgreSQL database
    Postgres,

    /// Oracle database
    Oracle,

    /// SQL Server database
    SqlServer,
}

impl SqlMetastoreDbType {
    /// Converts SQL placeholders to the database-specific format
    fn convert_placeholders(&self, sql: &str) -> String {
        match self {
            SqlMetastoreDbType::MySql => sql.to_string(),
            SqlMetastoreDbType::Postgres => {
                // Convert ? to $1, $2, etc.
                let re = Regex::new(r"\?").expect("Failed to create regex pattern for SQL placeholders");
                let mut counter = 0;
                re.replace_all(sql, |_: &regex::Captures<'_>| {
                    counter += 1;
                    format!("${}", counter)
                })
                .to_string()
            }
            SqlMetastoreDbType::Oracle => {
                // Convert ? to :1, :2, etc.
                let re = Regex::new(r"\?").expect("Failed to create regex pattern for SQL placeholders");
                let mut counter = 0;
                re.replace_all(sql, |_: &regex::Captures<'_>| {
                    counter += 1;
                    format!(":{}", counter)
                })
                .to_string()
            }
            SqlMetastoreDbType::SqlServer => {
                // Convert ? to @p1, @p2, etc.
                let re = Regex::new(r"\?").expect("Failed to create regex pattern for SQL placeholders");
                let mut counter = 0;
                re.replace_all(sql, |_: &regex::Captures<'_>| {
                    counter += 1;
                    format!("@p{}", counter)
                })
                .to_string()
            }
        }
    }
}

/// SQL database client trait for metastore operations
#[async_trait]
pub trait SqlClient: Send + Sync + std::fmt::Debug {
    /// Loads a key record by ID and created timestamp
    async fn load_key(
        &self,
        query: &str,
        id: &str,
        created: DateTime<Utc>,
    ) -> Result<Option<String>>;

    /// Loads the latest key record by ID
    async fn load_latest_key(&self, query: &str, id: &str) -> Result<Option<String>>;

    /// Stores a key record
    async fn store_key(
        &self,
        query: &str,
        id: &str,
        created: DateTime<Utc>,
        key_record: &str,
    ) -> Result<bool>;
}

/// SQL metastore implementation
#[derive(Debug)]
pub struct SqlMetastore {
    /// SQL client
    client: Arc<dyn SqlClient>,

    /// Database type (for future use)
    #[allow(dead_code)]
    db_type: SqlMetastoreDbType,

    /// Query for loading a key
    load_key_query: String,

    /// Query for storing a key
    store_key_query: String,

    /// Query for loading the latest key
    load_latest_query: String,
}

impl SqlMetastore {
    /// Creates a new SQL metastore with the given client and options
    pub fn new(client: Arc<dyn SqlClient>, db_type: SqlMetastoreDbType) -> Self {
        let load_key_query = db_type.convert_placeholders(DEFAULT_LOAD_KEY_QUERY);
        let store_key_query = db_type.convert_placeholders(DEFAULT_STORE_KEY_QUERY);
        let load_latest_query = db_type.convert_placeholders(DEFAULT_LOAD_LATEST_QUERY);

        Self {
            client,
            db_type,
            load_key_query,
            store_key_query,
            load_latest_query,
        }
    }

    /// Creates a new SQL metastore with the given client and custom queries
    pub fn with_custom_queries(
        client: Arc<dyn SqlClient>,
        db_type: SqlMetastoreDbType,
        load_key_query: impl Into<String>,
        store_key_query: impl Into<String>,
        load_latest_query: impl Into<String>,
    ) -> Self {
        let load_key_query = db_type.convert_placeholders(&load_key_query.into());
        let store_key_query = db_type.convert_placeholders(&store_key_query.into());
        let load_latest_query = db_type.convert_placeholders(&load_latest_query.into());

        Self {
            client,
            db_type,
            load_key_query,
            store_key_query,
            load_latest_query,
        }
    }

    /// Parses an envelope key record from a JSON string
    fn parse_envelope(json_str: &str) -> Result<EnvelopeKeyRecord> {
        serde_json::from_str(json_str)
            .map_err(|e| Error::Metastore(format!("Unable to parse key: {}", e)))
    }
}

#[async_trait]
impl Metastore for SqlMetastore {
    async fn load(&self, id: &str, created: i64) -> Result<Option<EnvelopeKeyRecord>> {
        let start = Instant::now();

        let created_dt = Utc
            .timestamp_opt(created, 0)
            .single()
            .ok_or_else(|| Error::Metastore("Invalid timestamp".to_string()))?;

        let result = match self
            .client
            .load_key(&self.load_key_query, id, created_dt)
            .await?
        {
            Some(json_str) => Some(Self::parse_envelope(&json_str)?),
            None => None,
        };

        histogram!("ael.metastore.sql.load", start.elapsed());

        Ok(result)
    }

    async fn load_latest(&self, id: &str) -> Result<Option<EnvelopeKeyRecord>> {
        let start = Instant::now();

        let result = match self
            .client
            .load_latest_key(&self.load_latest_query, id)
            .await?
        {
            Some(json_str) => Some(Self::parse_envelope(&json_str)?),
            None => None,
        };

        histogram!("ael.metastore.sql.loadlatest", start.elapsed());

        Ok(result)
    }

    async fn store(&self, id: &str, created: i64, envelope: &EnvelopeKeyRecord) -> Result<bool> {
        let start = Instant::now();

        let created_dt = Utc
            .timestamp_opt(created, 0)
            .single()
            .ok_or_else(|| Error::Metastore("Invalid timestamp".to_string()))?;

        let json_str = serde_json::to_string(envelope)
            .map_err(|e| Error::Metastore(format!("Error marshaling envelope: {}", e)))?;

        let result = self
            .client
            .store_key(&self.store_key_query, id, created_dt, &json_str)
            .await;

        histogram!("ael.metastore.sql.store", start.elapsed());

        result
    }
}
